srml/session/src/lib.rs

```diff
@@ -115,208 +115,306 @@
+				let maybe_old_key = old_keys.as_ref().map(|o| o.get_raw(i));
+				if maybe_old_key == Some(new_key) {
+					// no change.
+					updates.push(None);
+					continue;
+				}
+				let mut active = <Active<T>>::get(i as u32);
+				match active.binary_search_by(|k| k[..].cmp(&new_key)) {
+					Ok(_) => return Err("duplicate key provided"),
+					Err(pos) => active.insert(pos, new_key.to_owned()),
+				}
+				if let Some(old_key) = maybe_old_key {
+					match active.binary_search_by(|k| k[..].cmp(&old_key)) {
+						Ok(pos) => { active.remove(pos); }
+						Err(_) => {
+							// unreachable as long as our state is valid. we don't want to panic if
+							// it isn't, though.
+						}
+					}
+				}
+				updates.push(Some((i, active)));
+			}
 
-	/// Hook to be called after transaction processing.
-	pub fn check_rotate_session(block_number: T::BlockNumber) {
-		// Do this last, after the staking system has had the chance to switch out the authorities for the
-		// new set.
-		// Check block number and call `rotate_session` if necessary.
-		let is_final_block = ((block_number - Self::last_length_change()) % Self::length()).is_zero();
-		let (should_end_session, apply_rewards) = <ForcingNewSession<T>>::take()
-			.map_or((is_final_block, is_final_block), |apply_rewards| (true, apply_rewards));
-		if should_end_session {
-			Self::rotate_session(is_final_block, apply_rewards);
+			// Update the active sets.
+			for (i, active) in updates.into_iter().filter_map(|x| x) {
+				<Active<T>>::insert(i as u32, active);
+			}
+			// Set new keys value for next session.
+			<NextKeyFor<T>>::insert(who, keys);
+			// Something changed.
+			<Changed<T>>::put(true);
+		}
+
+		/// Called when a block is finalized. Will rotate session if it is the last
+		/// block of the current session.
+		fn on_initialize(n: T::BlockNumber) {
+			if T::ShouldEndSession::should_end_session(n) {
+				Self::rotate_session();
+			}
 		}
 	}
+}
 
+impl<T: Trait> Module<T> {
 	/// Move on to next session: register the new authority set.
-	pub fn rotate_session(is_final_block: bool, apply_rewards: bool) {
-		let now = <timestamp::Module<T>>::get();
-		let time_elapsed = now.clone() - Self::current_start();
-		let session_index = <CurrentIndex<T>>::get() + One::one();
-
-		Self::deposit_event(RawEvent::NewSession(session_index));
-
+	pub fn rotate_session() {
 		// Increment current session index.
-		<CurrentIndex<T>>::put(session_index);
-		<CurrentStart<T>>::put(now);
+		let session_index = <CurrentIndex<T>>::get();
+
+		let mut changed = <Changed<T>>::take();
 
-		// Enact session length change.
-		let len_changed = if let Some(next_len) = <NextSessionLength<T>>::take() {
-			<SessionLength<T>>::put(next_len);
-			true
+		// See if we have a new validator set.
+		let validators = if let Some(new) = T::OnSessionEnding::on_session_ending(session_index) {
+			changed = true;
+			<Validators<T>>::put(&new);
+			new
 		} else {
-			false
+			<Validators<T>>::get()
```

Comment:
`changed` will still be passed to `on_new_session` even if `OnSessionEnding` returns false.

Reply:
correct. what's the problem?

Reply:
It seems wrong. How can BABE tell `session` "please don't change the session! my epoch isn't over!"?

Reply:
Assuming it is wired correctly then Babe is the *only thing that can change the session*.